internal/fpath/path.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package fpath
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+const storjScheme = "sj"
+
+// FPath is an OS independently path handling structure
+type FPath struct {
+	original string // the original URL or local path
+	local    bool   // if local path
+	bucket   string // only for Storj URL
+	path     string // only for Storj URL - the path within the bucket, cleaned from duplicated slashes
+}
+
+// New creates new FPath from the given URL
+func New(p string) (FPath, error) {
+	fp := FPath{original: p}
+
+	if filepath.IsAbs(p) {
+		fp.local = true
+		return fp, nil
+	}
+
+Parse:
+	u, err := url.Parse(p)
```

Comment:
Use an infinite loop to make the intent more clear.

```
for p != "" {
   ... 
   if u.Host == "" && u.Path != "" {
       p = strings.Replace(p, ":///", "://", 1)
       continue
   }
   break
```

Reply:
Done. Although I used `for {` instead of `for p != ""`. The check was not necessary and caused problem if the input is empty string.